Unable to evaluate long tuple items in EE

At the breakpoint, evaluate `x.Item8`:

```
class C
{
    static void Main()
    {
        var x = (1, 2, 3, 4, 5, 6, 7, 8);
        System.Diagnostics.Debugger.Break();
    }
}
```

```
x.Item8
error CS1061: 'ValueTuple<int, int, int, int, int, int, int, (int)>' does not contain a definition for 'Item8' and no extension method 'Item8' accepting a first argument of type 'ValueTuple<int, int, int, int, int, int, int, (int)>' could be found (are you missing a using directive or an assembly reference?)
```

--- a/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpEESymbolProvider.cs
+++ b/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpEESymbolProvider.cs
@@ -108,11 +108,7 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
             {
                 type = DynamicTypeDecoder.TransformTypeWithoutCustomModifierFlags(type, _sourceAssembly, refKind, dynamicFlagsOpt, checkLength: false);
             }
-            if (!tupleElementNamesOpt.IsDefault)
-            {
-                type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNamesOpt);
-            }
-            return type;
+            return TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNamesOpt);
         }
     }
 }

--- a/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/PlaceholderLocalSymbol.cs
+++ b/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/PlaceholderLocalSymbol.cs
@@ -50,10 +50,7 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
                 type = DecodeDynamicTypes(type, sourceAssembly, dynamicFlags);
             }
 
-            if (tupleElementNames != null)
-            {
-                type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNames.AsImmutable());
-            }
+            type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNames.AsImmutableOrNull());
 
             var name = alias.FullName;
             var displayName = alias.Name;

--- a/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs
+++ b/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs
@@ -9,6 +9,7 @@ using Microsoft.VisualStudio.Debugger.Clr;
 using Microsoft.VisualStudio.Debugger.Evaluation;
 using Microsoft.VisualStudio.Debugger.Evaluation.ClrCompilation;
 using Roslyn.Test.Utilities;
+using System;
 using System.Collections.Immutable;
 using System.Collections.ObjectModel;
 using System.Linq;
@@ -51,7 +52,7 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator.UnitTests
 @"{
   // Code size        8 (0x8)
   .maxstack  2
-  .locals init (System.ValueTuple<int, int> V_0) //o
+  .locals init ((int, int) V_0) //o
   IL_0000:  ldc.i4.1
   IL_0001:  ldc.i4.2
   IL_0002:  newobj     ""System.ValueTuple<int, int>..ctor(int, int)""
@@ -103,7 +104,7 @@ class C
 @"{
   // Code size        8 (0x8)
   .maxstack  2
-  .locals init (System.ValueTuple<int, int> V_0) //o
+  .locals init ((int, int) V_0) //o
   IL_0000:  ldc.i4.1
   IL_0001:  ldc.i4.2
   IL_0002:  newobj     ""System.ValueTuple<int, int>..ctor(int, int)""
@@ -201,7 +202,77 @@ class C
                 locals.Free();
             });
         }
-        
+
+        [WorkItem(13803, "https://github.com/dotnet/roslyn/issues/13803")]
+        [Fact]
+        public void LongTupleLocalElement_NoNames()
+        {
+            var source =
+@"class C
+{
+    static void M()
+    {
+        var x = (1, 2, 3, 4, 5, 6, 7, 8);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib(source, new[] { ValueTupleRef }, options: TestOptions.DebugDll);
+            WithRuntimeInstance(comp, new[] { MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef }, runtime =>
+            {
+                var context = CreateMethodContext(runtime, "C.M");
+                var testData = new CompilationTestData();
+                string error;
+                context.CompileExpression("x.Item4 + x.Item8", out error, testData);
+                testData.GetMethodData("<>x.<>m0").VerifyIL(
+@"{
+  // Code size       19 (0x13)
+  .maxstack  2
+  .locals init ((int, int, int, int, int, int, int, int) V_0) //x
+  IL_0000:  ldloc.0
+  IL_0001:  ldfld      ""int System.ValueTuple<int, int, int, int, int, int, int, (int)>.Item4""
+  IL_0006:  ldloc.0
+  IL_0007:  ldfld      ""(int) System.ValueTuple<int, int, int, int, int, int, int, (int)>.Rest""
+  IL_000c:  ldfld      ""int System.ValueTuple<int>.Item1""
+  IL_0011:  add
+  IL_0012:  ret
+}");
+            });
+        }
+
+        [Fact]
+        public void LongTupleLocalElement_Names()
+        {
+            var source =
+@"class C
+{
+    static void M()
+    {
+        var x = (1, 2, Three: 3, Four: 4, 5, 6, 7, Eight: 8);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib(source, new[] { ValueTupleRef }, options: TestOptions.DebugDll);
+            WithRuntimeInstance(comp, new[] { MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef }, runtime =>
+            {
+                var context = CreateMethodContext(runtime, "C.M");
+                var testData = new CompilationTestData();
+                string error;
+                context.CompileExpression("x.Item8 + x.Eight", out error, testData);
+                testData.GetMethodData("<>x.<>m0").VerifyIL(
+@"{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init ((int, int, int Three, int Four, int, int, int, int Eight) V_0) //x
+  IL_0000:  ldloc.0
+  IL_0001:  ldfld      ""(int) System.ValueTuple<int, int, int, int, int, int, int, (int)>.Rest""
+  IL_0006:  ldfld      ""int System.ValueTuple<int>.Item1""
+  IL_000b:  ldloc.0
+  IL_000c:  ldfld      ""(int) System.ValueTuple<int, int, int, int, int, int, int, (int)>.Rest""
+  IL_0011:  ldfld      ""int System.ValueTuple<int>.Item1""
+  IL_0016:  add
+  IL_0017:  ret
+}");
+            });
+        }
+
         [Fact]
         public void DeclareLocal()
         {
@@ -246,7 +317,7 @@ class C
 @"{
   // Code size       64 (0x40)
   .maxstack  6
-  .locals init (System.ValueTuple<int, int> V_0) //x
+  .locals init ((int, int) V_0) //x
   IL_0000:  ldtoken    ""System.ValueTuple<int, int>""
   IL_0005:  call       ""System.Type System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)""
   IL_000a:  ldstr      ""y""
@@ -329,6 +400,74 @@ class C
   IL_000f:  ret
 }");
                 locals.Free();
+            });
+        }
+
+        [WorkItem(13803, "https://github.com/dotnet/roslyn/issues/13803")]
+        [Fact]
+        public void AliasElement_NoNames()
+        {
+            var source =
+@"class C
+{
+    static (int, int) F;
+    static void M()
+    {
+    }
+}";
+            var comp = CreateCompilationWithMscorlib(source, new[] { ValueTupleRef }, options: TestOptions.DebugDll);
+            WithRuntimeInstance(comp, new[] { MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef }, runtime =>
+            {
+                var context = CreateMethodContext(runtime, "C.M");
+                var alias = new Alias(
+                    DkmClrAliasKind.Variable,
+                    "x",
+                    "x",
+                    "System.ValueTuple`8[" +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                        "[System.ValueTuple`2[" +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], " +
+                        "System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]], " +
+                    "System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51",
+                    Guid.Empty,
+                    null);
+                ResultProperties resultProperties;
+                string error;
+                ImmutableArray<AssemblyIdentity> missingAssemblyIdentities;
+                var testData = new CompilationTestData();
+                context.CompileExpression(
+                    "x.Item4 + x.Item8",
+                    DkmEvaluationFlags.TreatAsExpression,
+                    ImmutableArray.Create(alias),
+                    DebuggerDiagnosticFormatter.Instance,
+                    out resultProperties,
+                    out error,
+                    out missingAssemblyIdentities,
+                    null,
+                    testData);
+                testData.GetMethodData("<>x.<>m0").VerifyIL(
+@"{
+  // Code size       47 (0x2f)
+  .maxstack  2
+  IL_0000:  ldstr      ""x""
+  IL_0005:  call       ""object Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetObjectByAlias(string)""
+  IL_000a:  unbox.any  ""System.ValueTuple<int, int, int, int, int, int, int, (int, int)>""
+  IL_000f:  ldfld      ""int System.ValueTuple<int, int, int, int, int, int, int, (int, int)>.Item4""
+  IL_0014:  ldstr      ""x""
+  IL_0019:  call       ""object Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetObjectByAlias(string)""
+  IL_001e:  unbox.any  ""System.ValueTuple<int, int, int, int, int, int, int, (int, int)>""
+  IL_0023:  ldfld      ""(int, int) System.ValueTuple<int, int, int, int, int, int, int, (int, int)>.Rest""
+  IL_0028:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_002d:  add
+  IL_002e:  ret
+}");
             });
         }
     }

--- a/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/PlaceholderLocalSymbol.vb
+++ b/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/PlaceholderLocalSymbol.vb
@@ -37,10 +37,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
             Dim dynamicFlags As ReadOnlyCollection(Of Byte) = Nothing
             Dim tupleElementNames As ReadOnlyCollection(Of String) = Nothing
             CustomTypeInfo.Decode([alias].CustomTypeInfoId, [alias].CustomTypeInfo, dynamicFlags, tupleElementNames)
-
-            If tupleElementNames IsNot Nothing Then
-                type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNames.AsImmutable())
-            End If
+            type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNames.AsImmutableOrNull())
 
             Dim name = [alias].FullName
             Dim displayName = [alias].Name

--- a/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicEESymbolProvider.vb
+++ b/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicEESymbolProvider.vb
@@ -76,9 +76,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
         End Function
 
         Private Function IncludeTupleElementNamesIfAny(type As TypeSymbol, tupleElementNamesOpt As ImmutableArray(Of String)) As TypeSymbol
-            Return If(tupleElementNamesOpt.IsDefault,
-                type,
-                TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNamesOpt))
+            Return TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, tupleElementNamesOpt)
         End Function
 
     End Class

--- a/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/TupleTests.vb
+++ b/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/TupleTests.vb
@@ -54,7 +54,7 @@ End Class"
 "{
   // Code size        8 (0x8)
   .maxstack  2
-  .locals init (System.ValueTuple(Of Integer, Integer) V_0) //o
+  .locals init ((Integer, Integer) V_0) //o
   IL_0000:  ldc.i4.1
   IL_0001:  ldc.i4.2
   IL_0002:  newobj     ""Sub System.ValueTuple(Of Integer, Integer)..ctor(Integer, Integer)""
@@ -105,7 +105,7 @@ End Class"
 "{
   // Code size        8 (0x8)
   .maxstack  2
-  .locals init (System.ValueTuple(Of Integer, Integer) V_0) //o
+  .locals init ((Integer, Integer) V_0) //o
   IL_0000:  ldc.i4.1
   IL_0001:  ldc.i4.2
   IL_0002:  newobj     ""Sub System.ValueTuple(Of Integer, Integer)..ctor(Integer, Integer)""
@@ -204,6 +204,82 @@ class C
                 End Sub)
         End Sub
 
+        <WorkItem(13803, "https://github.com/dotnet/roslyn/issues/13803")>
+        <Fact>
+        Public Sub LongTupleLocalElement_NoNames()
+            Const source =
+"Class C
+    Shared Sub M()
+        Dim x = (1, 2, 3, 4, 5, 6, 7, 8)
+    End Sub
+End Class"
+            Dim comp = CreateCompilationWithMscorlib({source}, references:={ValueTupleRef}, options:=TestOptions.DebugDll)
+            WithRuntimeInstance(comp,
+                {MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef},
+                Sub(runtime)
+                    Dim context = CreateMethodContext(runtime, "C.M")
+                    Dim errorMessage As String = Nothing
+                    Dim testData = New CompilationTestData()
+                    context.CompileExpression(
+                        "x.Item4 + x.Item8",
+                        DkmEvaluationFlags.TreatAsExpression,
+                        NoAliases,
+                        errorMessage,
+                        testData)
+                    testData.GetMethodData("<>x.<>m0").VerifyIL(
+"{
+  // Code size       19 (0x13)
+  .maxstack  2
+  .locals init ((Integer, Integer, Integer, Integer, Integer, Integer, Integer, Integer) V_0) //x
+  IL_0000:  ldloc.0
+  IL_0001:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer)).Item4 As Integer""
+  IL_0006:  ldloc.0
+  IL_0007:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer)).Rest As (Integer)""
+  IL_000c:  ldfld      ""System.ValueTuple(Of Integer).Item1 As Integer""
+  IL_0011:  add.ovf
+  IL_0012:  ret
+}")
+                End Sub)
+        End Sub
+
+        <Fact>
+        Public Sub LongTupleLocalElement_Names()
+            Const source =
+"Class C
+    Shared Sub M()
+        Dim x = (1, 2, Three:=3, Four:=4, 5, 6, 7, Eight:=8)
+    End Sub
+End Class"
+            Dim comp = CreateCompilationWithMscorlib({source}, references:={ValueTupleRef}, options:=TestOptions.DebugDll)
+            WithRuntimeInstance(comp,
+                {MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef},
+                Sub(runtime)
+                    Dim context = CreateMethodContext(runtime, "C.M")
+                    Dim errorMessage As String = Nothing
+                    Dim testData = New CompilationTestData()
+                    context.CompileExpression(
+                        "x.Item8 + x.Eight",
+                        DkmEvaluationFlags.TreatAsExpression,
+                        NoAliases,
+                        errorMessage,
+                        testData)
+                    testData.GetMethodData("<>x.<>m0").VerifyIL(
+"{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init ((Integer, Integer, Three As Integer, Four As Integer, Integer, Integer, Integer, Eight As Integer) V_0) //x
+  IL_0000:  ldloc.0
+  IL_0001:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer)).Rest As (Integer)""
+  IL_0006:  ldfld      ""System.ValueTuple(Of Integer).Item1 As Integer""
+  IL_000b:  ldloc.0
+  IL_000c:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer)).Rest As (Integer)""
+  IL_0011:  ldfld      ""System.ValueTuple(Of Integer).Item1 As Integer""
+  IL_0016:  add.ovf
+  IL_0017:  ret
+}")
+                End Sub)
+        End Sub
+
         ''' <summary>
         ''' Locals declared in the VB EE do not have an explicit
         ''' type and are statically typed to Object, so tuple
@@ -240,7 +316,7 @@ End Class"
 "{
   // Code size       48 (0x30)
   .maxstack  4
-  .locals init (System.ValueTuple(Of Integer, Integer) V_0, //x
+  .locals init ((Integer, Integer) V_0, //x
                 System.Guid V_1)
   IL_0000:  ldtoken    ""Object""
   IL_0005:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
@@ -257,7 +333,7 @@ End Class"
   IL_002e:  stind.ref
   IL_002f:  ret
 }")
-                End Sub)
+                            End Sub)
         End Sub
 
         <WorkItem(13589, "https://github.com/dotnet/roslyn/issues/13589")>
@@ -316,6 +392,67 @@ End Class"
   IL_000f:  ret
 }")
                     locals.Free()
+                End Sub)
+        End Sub
+
+        <WorkItem(13803, "https://github.com/dotnet/roslyn/issues/13803")>
+        <Fact>
+        Public Sub AliasElement_NoNames()
+            Const source =
+"Class C
+    Shared F As (Integer, Integer)
+    Shared Sub M()
+    End Sub
+End Class"
+            Dim comp = CreateCompilationWithMscorlib({source}, references:={ValueTupleRef}, options:=TestOptions.DebugDll)
+            WithRuntimeInstance(comp,
+                {MscorlibRef, SystemCoreRef, SystemRuntimeFacadeRef, ValueTupleRef},
+                Sub(runtime)
+                    Dim context = CreateMethodContext(runtime, "C.M")
+                    Dim [alias] = New [Alias](
+                        DkmClrAliasKind.Variable,
+                        "x",
+                        "x",
+                        "System.ValueTuple`8[" +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                            "[System.ValueTuple`2[" +
+                                "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]," +
+                                "[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], " +
+                            "System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]], " +
+                        "System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51",
+                        Guid.Empty,
+                        Nothing)
+                    Dim errorMessage As String = Nothing
+                    Dim testData = New CompilationTestData()
+                    context.CompileExpression(
+                        "x.Item4 + x.Item8",
+                        DkmEvaluationFlags.TreatAsExpression,
+                        ImmutableArray.Create([alias]),
+                        errorMessage,
+                        testData)
+                    Assert.Null(errorMessage)
+                    testData.GetMethodData("<>x.<>m0").VerifyIL(
+"{
+  // Code size       47 (0x2f)
+  .maxstack  2
+  IL_0000:  ldstr      ""x""
+  IL_0005:  call       ""Function Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetObjectByAlias(String) As Object""
+  IL_000a:  unbox.any  ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer, Integer))""
+  IL_000f:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer, Integer)).Item4 As Integer""
+  IL_0014:  ldstr      ""x""
+  IL_0019:  call       ""Function Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetObjectByAlias(String) As Object""
+  IL_001e:  unbox.any  ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer, Integer))""
+  IL_0023:  ldfld      ""System.ValueTuple(Of Integer, Integer, Integer, Integer, Integer, Integer, Integer, (Integer, Integer)).Rest As (Integer, Integer)""
+  IL_0028:  ldfld      ""System.ValueTuple(Of Integer, Integer).Item1 As Integer""
+  IL_002d:  add.ovf
+  IL_002e:  ret
+}")
                 End Sub)
         End Sub
 

